src/assignment-10/FlashLoanLiquidator.sol

```diff
@@ -0,0 +1,123 @@
+import { CollateralizedVault } from "src/assignment-10/CollateralizedVault.sol";
+import { AMM } from "src/assignment-7/AMM.sol";
+import { TransferHelper } from "yield-utils-v2/token/TransferHelper.sol";
+
+/// @title Flash loan liquidator
+/// @author davidbrai
+/// @notice Uses a flash loan in order to liquidate an undercollateralized debt position and keep the arbitrage
+/// @dev Uses UniswapV2Pair for flash swaps
+contract FlashLoanLiquidator {
+
+    using TransferHelper for IERC20;
+
+    /******************
+     * Immutables
+     ******************/
+
+    IERC20 immutable public underlying;
+    IERC20 immutable public collateral;
+    IUniswapV2Factory immutable public uniswapFactory;
+    CollateralizedVault immutable public vault;
+    AMM immutable public amm;
+
+
+    /******************
+     * Storage
+     ******************/
+
+    /// @dev Security: this is used to make sure the callback is only called by the expected UniV2 pair contract
+    address public permissionedPair;
+
+    /******************
+     * Events
+     ******************/
+    event Liquidate(address indexed liquidator, address indexed liquidatee, uint256 profit);
+
+    /******************
+     * Errors
+     ******************/
+
+    error UnauthorizedInitiator();
+    error UnauthorizedMsgSender();
+
+    struct SwapParams {
+        address vaultUser;
+        address liquidator;
+    }
+
+    /// @notice Creates a new FlashLoanLiquidator
+    /// @param underlying_ ERC20 token address of the asset being borrowed
+    /// @param collateral_ ERC20 token address of the collateral asset
+    /// @param uniswapFactory_ address of UniswapV2Factory, used for finding UniswapV2Pair contracts to use for flash swaps
+    /// @param vault_ A CollateralizedVault contract against which to liquidate
+    /// @param amm_ An AMM for swapping the `collateral_` back into `underlying_`
+    constructor(address underlying_, address collateral_, address uniswapFactory_, address vault_, address amm_) {
+        underlying = IERC20(underlying_);
+        collateral = IERC20(collateral_);
+        uniswapFactory = IUniswapV2Factory(uniswapFactory_);
+        vault = CollateralizedVault(vault_);
+        amm = AMM(amm_);
+    }
+
+    /// @notice Takes a flash loan of `underlying` and liquidates an undercollateralized user debt position
+    ///     Then is swaps the `collateral` back into `underlying` tokens, repays the flash loan, and send
+    ///     the profit back to who invoked this function.
+    /// @dev UniswapV2Pair is used for flash loaning `underlying`
+    /// @param user The debt position to liquidate in `vault`
+    function liquidate(address user) public {
+        uint256 debt = vault.borrows(user);
+
+        // params we want to be available in the flash loan callback (`uniswapV2Call`)
+        bytes memory data = abi.encode(SwapParams({vaultUser: user, liquidator: msg.sender}));
```

Comment:
Now it's a good time to talk about the [dark forest](https://medium.com/@danrobinson/ethereum-is-a-dark-forest-ecc5f0505dff). Try to tell me why `msg.sender` will never see a profit.

Reply:
@alcueca if this is sent via flashbots, should be able to see some profit, no?

Reply:
if my smart contract is the most gas effecient, I should be able to offer the most miner bribe (block.coinbase) while still profiting from the little gas optimization I have compared to other searchers

Reply:
Ah, maybe, forget that flashbots is now the way to go.